### PR TITLE
DOC: MEP 28 Artist Refactor

### PR DESCRIPTION
**Pull request by OceanWolf**
Something perhaps for 2.1.  Something I can work on after MEP27 :wink:.

### CODE REVIEW

**Comment by tacaswell** [comment]
There is also an effort started at scipy2015 sprints by @rmorshea which is related to this.  I just opened a PR (#4694) from his branch.

**Comment by tacaswell** [comment]
You should also have a look at the `Container` related work in #3944

**Comment by OceanWolf** [comment]
Big question first, do you think this needs a MEP?  I started writing this as an issue and then it became rather long, so I thought I should convert it.  Looking at the scipy sprint I think we need to outline the direction, perhaps with some branches to show proof of concepts.

I took a look a brief look at #3944, a bit confused as to the difference between a ``Collection`` and a ``Container``, I see no documentation on the api for containers... http://matplotlib.org/api/ and ``containers.py`` has very little doc.  It doesn't look like a very large file, so I guess I will figure it out.

What timescale did you have in mind for this?  I see you marked this as "proposed next point release", does that relate to the MEP doc, or the code itself?  I imagine you don't want to get bothered with code now just two weeks before release as I imagine you have a lot of other stuff to work on.  I find it hard enough just keeping up with all the notifications, every time I check this last day I see another 30 new notifications, so I thought twice about posting this PR, but as an MEP discussion document I thought it could simply go on the table for a few weeks.

**Comment by efiring** [comment]
On 2015/07/14 8:48 AM, OceanWolf wrote:
> I took a look a brief look at #3944
> <https://github.com/matplotlib/matplotlib/pull/3944>, a bit confused as
> to the difference between a |Collection| and a |Container|, I see no
> documentation on the api for containers... http://matplotlib.org/api/
> and |containers.py| has very little doc. It doesn't look like a very
> large file, so I guess I will figure it out.

Collections are for multiple objects that are very similar; they are 
designed to take advantage of this similarity to allow backends to 
render them much more efficiently than would be the case for a sequence 
of individual objects.  Notice how backends override the base class 
rendering to gain efficiency.

Containers are for different types of object that combine to make a 
complicated plot type; and/or for combining objects that are the same 
type, but that don't easily go into a single collection.

**Comment by OceanWolf** [comment]
Does matplotlib have a list of both ancestor and descendant classes, especially the latter?  Something like [this](http://lazka.github.io/pgi-docs/Gtk-3.0/classes/Widget.html#Gtk.Widget) which lists the subclasses?

**Comment by tacaswell** [comment]
see http://matplotlib.org/api/artist_api.html

File: doc/devel/MEP/MEP28.rst [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+All development branches containing work on this MEP should be linked to from here.
+
+All pull requests submitted relating to this MEP should be linked to
+from here.  (A MEP does not need to be implemented in a single pull
+request if it makes sense to implement it in discrete phases).
+
+Abstract
+========
+
+During the course of MPL, the number and complexity of artists have grown, to become quite unmanageable, this PR seeks to address that by refactoring the Artists and creating a more sensible class structure.
```
Comment (efiring) [inline doc/devel/MEP/MEP28.rst]: I think all of our rst files should use hard returns to keep line lengths reasonable.  I haven't checked, but I assume this has been done in the past.

File: doc/devel/MEP/MEP28.rst [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+With a clearer structure it also makes it easier to add new features such as keeping the legend upto date with its parent artists, add make it easier to implement dynamical property changing.  
+
+
+Implementation
+==============
+
+## Step 1 - Remove getter/setters
+The easiest and biggest reduction comes from removing all of the ``artist.set_xxx(value)`` methods (where possible).
+We will maintain a list of the properties the Artist class held as a class property.
+To maintain BC we will add a single ``__getattr__`` in the Artist base class to return a closure to set a dict property.
```
Comment (efiring) [inline doc/devel/MEP/MEP28.rst]: I don't understand what you mean here.

File: doc/devel/MEP/MEP28.rst [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+The goal of this MEP lies in automating and reducing as much of the "admin" code as possible making it easier to see what goes on and where and make faster changes to the codebase.  A lot of the artist code for the example gets taken up with simple getter/setters, very laborious, and when it comes to the documentation it looks like a getter/setter desert. More so, we the same properties duplicated throughout the codebase, such as fillcolor, where we have ``set_fillcolor`` methods in ``lines``, ``markers`` and ``patches``.  This makes readability and maintainability very difficult and just adds to the Artist doc desert.
+
+With a clearer structure it also makes it easier to add new features such as keeping the legend upto date with its parent artists, add make it easier to implement dynamical property changing.  
+
+
+Implementation
+==============
+
+## Step 1 - Remove getter/setters
+The easiest and biggest reduction comes from removing all of the ``artist.set_xxx(value)`` methods (where possible).
```
Comment (efiring) [inline doc/devel/MEP/MEP28.rst]: By using plain attributes, or by using properties?  With the latter, we will still have getters and setters, but not as part of the API.

File: doc/devel/MEP/MEP28.rst [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+We will maintain a list of the properties the Artist class held as a class property.
+To maintain BC we will add a single ``__getattr__`` in the Artist base class to return a closure to set a dict property.
+
+With this done, we can easily create property getter/setters, i.e. ``artist.xxx = value`` with  ``__setattr__``.
```
Comment (efiring) [inline doc/devel/MEP/MEP28.rst]: It sounds like you are talking about packing a lot into __getattr__ and __setattr__, but I probably won't understand until I see an example.

**Comment by jklymak** [comment]
Not even clear we are doing MEPs anymore.  I'll close but feel free to reopen if this is still being pushed somewhere...

**Pull request closed without merging**